src/object_ref.jl

```diff
@@ -2,10 +2,33 @@ mutable struct ObjectRef
     oid::ray_jll.ObjectIDAllocated
     owner_address::Union{ray_jll.AddressAllocated,Nothing}
     serialized_object_status::String
+
+    function ObjectRef(oid, owner_address, serialized_object_status;
+                       add_local_ref=true)
+        if add_local_ref
+            worker = ray_jll.GetCoreWorker()
+            ray_jll.AddLocalReference(worker, oid)
+        end
+        objref = new(oid, owner_address, serialized_object_status)
+        return finalizer(objref) do objref
+            # putting finalizer behind `@async` may not be necessary since docs
+            # suggest that you should `ccall` IO functions.  But doing it this
+            # way allows us to do things like debug logging...
+            errormonitor(@async finalize_object_ref(objref))
+            return nothing
+        end
```

Comment:
Do you want to conditionally add the `finalizer` based upon `add_local_ref`? Also, since you can add `finalizer`s outside of the constructor it may nicer to separate reference tracking hooks from creating `ObjectRef` instances.

Reply:
(gonna respond to both questions here since they're good ones and related IMO)

I _think_ the only reason to _not_ call `AddLocalReference` during construction is when you get the ID from a core worker operation that _also_ increments the local ref; however, in that case, you _still_ want to _decrement_ the ref when the instance is GCed IIUC.

Reply:
for instance, the object IDs created during task submission are initialized with ref count 1, but the doc string says that the frontend is still responsible for decrementing them when they go out of scope in the application:

```
  /// Add a task that is pending execution.
  ///
  /// The local ref count for all return refs (excluding actor creation tasks)
  /// will be initialized to 1 so that the ref is considered in scope before
  /// returning to the language frontend. The caller is responsible for
  /// decrementing the ref count once the frontend ref has gone out of scope.
  ///
  /// \param[in] caller_address The rpc address of the calling task.
  /// \param[in] spec The spec of the pending task.
  /// \param[in] max_retries Number of times this task may be retried
  /// on failure.
  /// \return ObjectRefs returned by this task.
  std::vector<rpc::ObjectReference> AddPendingTask(const rpc::Address &caller_address,
                                                   const TaskSpecification &spec,
                                                   const std::string &call_site,
                                                   int max_retries = 0);
```

(it's a bit unfortunate that teh `CoreWorker::SubmitTask` docstring does _not_ mention this, you have to dig a bit into [the task manager](https://github.com/beacon-biosignals/ray/blob/4ceb62daaad05124713ff9d94ffbdad35ee19f86/src/ray/core_worker/task_manager.h#L199-L202))

Reply:
Another case: `CoreWorker::Put` calls `AddOwnedReference` and sets the `add_local_ref` arg to `true`:

```
Status CoreWorker::Put(const RayObject &object,
                       const std::vector<ObjectID> &contained_object_ids,
                       ObjectID *object_id) {
  *object_id = ObjectID::FromIndex(worker_context_.GetCurrentInternalTaskId(),
                                   worker_context_.GetNextPutIndex());
  reference_counter_->AddOwnedObject(*object_id,
                                     contained_object_ids,
                                     rpc_address_,
                                     CurrentCallSite(),
                                     object.GetSize(),
                                     /*is_reconstructable=*/false,
                                     /*add_local_ref=*/true,
                                     NodeID::FromBinary(rpc_address_.raylet_id()));
```
https://github.com/beacon-biosignals/ray/blob/4ceb62daaad05124713ff9d94ffbdad35ee19f86/src/ray/core_worker/core_worker.cc#L1110-L1122

```
  /// Add an object that we own. The object may depend on other objects.
  /// Dependencies for each ObjectID must be set at most once. The local
  /// reference count for the ObjectID is set to zero, which assumes that an
  /// ObjectID for it will be created in the language frontend after this call.
  ///
  /// TODO(swang): We could avoid copying the owner_address since
  /// we are the owner, but it is easier to store a copy for now, since the
  /// owner ID will change for workers executing normal tasks and it is
  /// possible to have leftover references after a task has finished.
  ///
  /// \param[in] object_id The ID of the object that we own.
  /// \param[in] contained_ids ObjectIDs that are contained in the object's value.
  /// As long as the object_id is in scope, the inner objects should not be GC'ed.
  /// \param[in] owner_address The address of the object's owner.
  /// \param[in] call_site Description of the call site where the reference was created.
  /// \param[in] object_size Object size if known, otherwise -1;
  /// \param[in] is_reconstructable Whether the object can be reconstructed
  /// through lineage re-execution.
  /// \param[in] add_local_ref Whether to initialize the local ref count to 1.
  /// This is used to ensure that the ref is considered in scope before the
  /// corresponding ObjectRef has been returned to the language frontend.
  /// \param[in] pinned_at_raylet_id The primary location for the object, if it
  /// is already known. This is only used for ray.put calls.
  void AddOwnedObject(const ObjectID &object_id,
                      const std::vector<ObjectID> &contained_ids,
                      const rpc::Address &owner_address,
                      const std::string &call_site,
                      const int64_t object_size,
                      bool is_reconstructable,
                      bool add_local_ref,
                      const absl::optional<NodeID> &pinned_at_raylet_id =
                          absl::optional<NodeID>()) LOCKS_EXCLUDED(mutex_);
```
https://github.com/beacon-biosignals/ray/blob/4ceb62daaad05124713ff9d94ffbdad35ee19f86/src/ray/core_worker/reference_count.h#L160-L163

Reply:
in _both_ those cases, we need the finalizer to remove the local ref even though we don't manually add the local ref when that object is created.  I think that, in general, _every_ instance of `ObjectRef` needs to "clean up" after itself like this, but not all of them need to add a local reference on initialization.  So, should we move the add local ref out of hte constructor?  I don't think so, because we don't necessarily control every code path by which an ObjectRef might be created (i.e., copy), so for safety's sake I think its best to default to always adding a local ref on construction.